src/sctools/fastq.py

```diff
@@ -160,41 +160,83 @@ def __iter__(self):
         """
         parse fastq files for barcodes(s) defined by EmbeddedBarcode objects, producing a generator
         that yields extracted barcode objects in a form that is consumable by pysam's bam and sam
         set_tag methods.
 
         :param [EmbeddedBarcode] embedded_barcodes: list of tag objects defining start and end of
           the sequence containing the tag, plus the string quality and sequence tags
-        :param list|str files: (default sys.stdin) fastq file or list of files to be read.
+        :param list|str fastq_files: (default sys.stdin) fastq file or list of files to be read.
         :param mode: (Default 'r') will return string objects. Change to 'rb' to
           return bytes objects.
         """
-        super().__init__(*args, **kwargs)
+        super().__init__(files=fastq_files, *args, **kwargs)
         self.embedded_barcodes = embedded_barcodes
 
     def __iter__(self):
         """iterates over barcodes extracted from fastq"""
         for record in super().__iter__():  # iterates records; we extract barcodes.
             barcodes = []
             for barcode in self.embedded_barcodes:
-                barcodes.extend(self.extract_barcode(record, barcode))
+                barcodes.extend(extract_barcode(record, barcode))
             yield barcodes
 
-    @staticmethod
-    def extract_barcode(record, eb):
-        """extracts barcodes from a fastq record according to an EmbeddedBarcode object.
-
-        :param FastqRecord record: record to extract from
-        :param EmbeddedBarcode eb: defines the barcode start and end and the names of the quality
-          and sequence tags
-        :return tuple: (sequence tag name, tag content (nucleotides), tag type (Z))
-        :return tuple: (quality tag name, tag content (qualities), tag type (Z))
+
+class BarcodeGeneratorWithCorrectedCellBarcodes(Reader):
+
+    def __init__(self, fastq_files, embedded_cell_barcode, whitelist, *other_embedded_barcodes):
         """
-        seq = record.sequence[eb.start:eb.end]
-        qual = record.quality[eb.start:eb.end]
-        return (eb.sequence_tag, seq, 'Z'), (eb.quality_tag, qual, 'Z')
+        parse fastq files for barcodes(s) defined by EmbeddedBarcode objects, producing a generator
+        that yields extracted barcode objects in a form that is consumable by pysam's bam and sam
+        set_tag methods.
+
+        :param list|str fastq_files: (default sys.stdin) fastq file or list of files to be read.
+        :param EmbeddedBarcode embedded_cell_barcode: cell barcode tag
+        :param [EmbeddedBarcode] other_embedded_barcodes: list of tag objects defining start and
+          end of the sequence containing the tag, plus the string quality and sequence tags
+        :param str whitelist: whitelist file containing "correct" cell barcodes for an experiment
+        :param mode: (Default 'r') will return string objects. Change to 'rb' to
```

Comment:
Is the `mode` parameter getting used anywhere?

Reply:
Yep! I had forgotten to include *args and **kwargs in init; it gets passed to the super class (reader) if it's present.